### PR TITLE
[android] Stop leaking FDs in parent test process.

### PR DESCRIPTION
**Pull request by drodriguez**
In the Android paths of the spawnChild function, the parent was creating
a pipe that was never closed, which led to FD starvation. In some tests
with a lots of expected crashes, the childs will not spawn anymore since
the linker would not have enough descriptors to open the shared
libraries, while in other tests which closed the child descriptors as
part of the last test, the parent process will hang waiting those
descriptors to be closed, which will never had happened.

The solution is implement the missing parts of the code, which tried to
read from the pipe in the parent side (using select and read, taking
pieces from other parts of the code). This should match the fork/execv
path used by Android and Haiku to the spawn code used by the rest of the
platforms.

This change fixes StdlibUnittest/Stdin.swift,
stdlib/InputStream.swift.gyb,
stdlib/Collection/FlattenCollection.swift.gyb and
stdlib/Collection/LazyFilterCollection.swift.gyb, which were the last 4
tests failing in Android AArch64.

/cc @zhuowei which wrote the original code and @return since this code might affect Haiku

### CODE REVIEW

File: stdlib/private/SwiftPrivateLibcExtras/Subprocess.swift [hunk 1]
```diff
@@ -261,6 +265,41 @@ public func spawnChild(_ args: [String])
...
+    var writefds = _stdlib_fd_set()
+    var errorfds = _stdlib_fd_set()
+    errorfds.set(childToParentPipe.readFD)
+
+    var ret: CInt
+    repeat {
+      ret = _stdlib_select(&readfds, &writefds, &errorfds, nil)
+    } while ret == -1 && errno == EINTR
+    if ret <= 0 {
+      fatalError("select() returned an error")
```
Comment (compnerd) [inline stdlib/private/SwiftPrivateLibcExtras/Subprocess.swift]: please print `errno` here.

File: stdlib/private/SwiftPrivateLibcExtras/Subprocess.swift [hunk 2]
```diff
@@ -261,6 +265,41 @@ public func spawnChild(_ args: [String])
 
     // Close the pipe when we're done writing the error.
     close(childToParentPipe.writeFD)
+  } else {
+    close(childToParentPipe.writeFD)
+
+    // Figure out if the child was execve correctly or not.
```
Comment (compnerd) [inline stdlib/private/SwiftPrivateLibcExtras/Subprocess.swift]: This comment doesn't make sense to me.

File: stdlib/private/SwiftPrivateLibcExtras/Subprocess.swift [hunk 3]
```diff
@@ -261,6 +265,41 @@ public func spawnChild(_ args: [String])
...
+      ret = _stdlib_select(&readfds, &writefds, &errorfds, nil)
+    } while ret == -1 && errno == EINTR
+    if ret <= 0 {
+      fatalError("select() returned an error")
+    }
+
+    if readfds.isset(childToParentPipe.readFD) || errorfds.isset(childToParentPipe.readFD) {
+      var childErrno: CInt = 0
+      let readResult: ssize_t = withUnsafeMutablePointer(to: &childErrno) {
+        return read(childToParentPipe.readFD, $0, MemoryLayout<CInt>.size)
```
Comment (compnerd) [inline stdlib/private/SwiftPrivateLibcExtras/Subprocess.swift]: Can you make this `MemoryLayout(of: childErrno).size` please?

File: stdlib/private/SwiftPrivateLibcExtras/Subprocess.swift [hunk 4]
```diff
@@ -261,6 +265,41 @@ public func spawnChild(_ args: [String])
...
+      }
+      if readResult == 0 {
+        // We found EOF, which happens when child execve is successful.
```
Comment (compnerd) [inline stdlib/private/SwiftPrivateLibcExtras/Subprocess.swift]: I think you missed some word here.

`We read an `EOF` indicating that the child was exec'ed successfully.`

File: stdlib/private/SwiftPrivateLibcExtras/Subprocess.swift [hunk 5]
```diff
@@ -261,6 +265,41 @@ public func spawnChild(_ args: [String])
...
+      } else if readResult < 0 {
+        fatalError("read() returned error")
```
Comment (compnerd) [inline stdlib/private/SwiftPrivateLibcExtras/Subprocess.swift]: Please print `errno` here.

File: stdlib/private/SwiftPrivateLibcExtras/Subprocess.swift (hunk 2 above)
Comment (drodriguez) [reply stdlib/private/SwiftPrivateLibcExtras/Subprocess.swift]: I think I was bending English a little bit too much in these comments. I was using “execve” as a verb. I wanted to say “Figure out if the child’s call to execve was successful or not”. I will change it.

**Comment by drodriguez** [comment]
Updated incorporating the provided feedback.

@swift-ci please test

**Comment by buttaface** [comment]
Tested this pull with the latest 5/12 snapshot and can confirm that it fixes those four tests on Android AArch64, the first two of which I've always disabled because they'd always hang.

**Review by compnerd — APPROVED** [review]

**Pull request merged**